### PR TITLE
add target assign operator for supporting retinanet

### PR DESCRIPTION
**Pull request by FlyingQianMM**
test=develop

This operator can be, for given the Intersection-over-Union (IoU) overlap between anchors and ground truth boxes, to assign classification and regression targets to each anchor, these target labels are used for training retinanet. No other sampling strategy is adopted to filter negative anchors because all negative labels along with positive labels contribute to the classfication objective(focal loss).
image

### CODE REVIEW

**Review by heavengate — COMMENTED** [review]
加一下paddle/fluid/API.spec, 在python/paddle/fluid/tests/unittests/test_layers里面加一下python API的单测， 另外涉及API的改动现在都需要生成文档预览张贴在pr里

File: paddle/fluid/operators/detection/retinanet_target_assign_op.cc [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+    PADDLE_ENFORCE_EQ(anchor_dims.size(), 2,
+                      "The rank of Input(Anchor) must be 2.");
+    PADDLE_ENFORCE_EQ(gt_boxes_dims.size(), 2,
+                      "The rank of Input(GtBoxes) must be 2.");
+    PADDLE_ENFORCE_EQ(gt_labels_dims.size(), 2,
+                      "The rank of Input(GtLabels) must be 2.");
+    PADDLE_ENFORCE_EQ(im_info_dims.size(), 2,
+                      "The rank of Input(ImInfo) must be 2.");
+
+    ctx->SetOutputDim("LocationIndex", {-1});
```
Comment (heavengate) [inline paddle/fluid/operators/detection/retinanet_target_assign_op.cc]: 这里不要用-1，用gt_boxes_dims[0]

File: paddle/fluid/operators/detection/retinanet_target_assign_op.cc [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+    auto* tgt_lbl = context.Output<LoDTensor>("TargetLabel");
+    auto* bbox_inside_weight = context.Output<LoDTensor>("BBoxInsideWeight");
+    auto* fg_num = context.Output<LoDTensor>("ForegroundNumber");
+
+    PADDLE_ENFORCE_EQ(gt_boxes->lod().size(), 1UL,
+                      "RetinanetTargetAssignOp gt_boxes needs 1 level of LoD");
+    PADDLE_ENFORCE_EQ(gt_labels->lod().size(), 1UL,
+                      "RetinanetTargetAssignOp gt_boxes needs 1 level of LoD");
+    PADDLE_ENFORCE_EQ(is_crowd->lod().size(), 1UL,
+                      "RetinanetTargetAssignOp is_crowd needs 1 level of LoD");
```
Comment (heavengate) [inline paddle/fluid/operators/detection/retinanet_target_assign_op.cc]: 这些最好在infershape里做

File: paddle/fluid/operators/detection/retinanet_target_assign_op.cc [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+    AddComment(R"DOC(
+    This layer can be, for given the Intersection-over-Union (IoU) overlap
+    between anchors and ground truth boxes, to assign classification and
+    regression targets to each each anchor, these target labels are used for
+    train retinanet. 
+    
+    Every anchor is assigned with a length num_classes
+    one-hot vector of classification targets, and a 4-vector of box regression
+    targets, where num_classes is the class number. The positive anchors are
+    two kinds of anchors: (i) the anchor/anchors with the highest IoU overlap
```
Comment (heavengate) [inline paddle/fluid/operators/detection/retinanet_target_assign_op.cc]: 给些公式描述，(i)这种分一下段，1. 2. 3.每个一段

File: python/paddle/fluid/layers/detection.py [hunk 4]
```diff
@@ -58,6 +59,155 @@
...
+                            cls_logits,
+                            anchor_box,
+                            anchor_var,
+                            gt_boxes,
+                            gt_labels,
+                            is_crowd,
+                            im_info,
+                            num_classes=1,
+                            positive_overlap=0.5,
+                            negative_overlap=0.4):
```
Comment (heavengate) [inline python/paddle/fluid/layers/detection.py:71]: add param ‘name'

**Review by heavengate — DISMISSED** [review]

File: paddle/fluid/operators/detection/retinanet_target_assign_op.cc (hunk 2 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/retinanet_target_assign_op.cc]: infershape里面无法取出输入tensor的指针，所以没能在infershape里判断lod信息

File: paddle/fluid/operators/detection/retinanet_target_assign_op.cc (hunk 1 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/retinanet_target_assign_op.cc]: 已修改

File: paddle/fluid/operators/detection/retinanet_target_assign_op.cc (hunk 3 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/retinanet_target_assign_op.cc]: 已修改

File: python/paddle/fluid/layers/detection.py (hunk 4 above)
Comment (FlyingQianMM) [reply python/paddle/fluid/layers/detection.py:71]: 因为op的输出有多个，此时一个name没法用来定义多个输出，因此省略掉了name

**Review by heavengate — DISMISSED** [review]

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 5]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
   }
 };
 
+class RetinanetTargetAssignOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Anchor",
+             "(Tensor) input anchor is a 2-D Tensor with shape [H*W*A, 4].");
+    AddInput("GtBoxes",
+             "(LoDTensor) input groud-truth bbox with shape [K, 4].");
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: ground-truth

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 6]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+    AddInput("GtLabels",
+             "(LoDTensor) input groud-truth label with shape [K, C]. "
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: 同上

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 7]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+             "C is the class number.");
+    AddInput("IsCrowd",
+             "(LoDTensor) input which indicates groud-truth is crowd.");
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: 同上

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 8]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+    AddOutput(
+        "LocationIndex",
+        "(Tensor), The indexes of foreground anchors in all anchors, the "
+        "shape of the LocationIndex is [F], F depends on the value of input "
+        "tensor and attributes.");
+    AddOutput(
+        "ScoreIndex",
+        "(Tensor), The indexes of foreground and background anchors in all "
+        "RPN anchors(The rest anchors are ignored). The shape of the "
+        "ScoreIndex is [F + B], F and B are foreground and backgroud "
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: background

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 9]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+        " number.");
+    AddOutput("TargetBBox",
+              "(Tensor), The target bbox deltas with shape "
+              "[F, 4], F is the foreground number.");
+    AddOutput("TargetLabel",
+              "(Tensor<int>), The target labels of each anchor with shape "
+              "[F + B, 1], F and B are foreground and backgroud number.");
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: background

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 10]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+    where C is the class number. The assignment rules are as followed:
+    
+    1. Anchors are assigned to ground-truth boxes when: (i) it has the highest
+    IoU overlap with a ground-truth box, or (ii) it has an IoU overlap higher
+    than positive_overlap(0.5) with any ground-truth box.
+    
+    2. Anchors are assigned to background when its IoU ratio is lower than
+    negative_overlap (0.4) for all ground-truth boxes.
+
+    When a anchor is assgned with a ground-truth box which is the i-th category,
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: assigned

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 11]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+    the i-th entry in its C vector of targets is set to 1 and all other entries
+    are set to 0. When a anchor is assgned with background, all entries are set
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: 同上

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 12]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+    to 0. Anchors that are not assigned do not contribute to the training
+    objective. The regression targets are the encoded ground-truth boxes
+    associated with the assgined anchors.
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: 同上

File: python/paddle/fluid/layers/detection.py [hunk 13]
```diff
@@ -58,6 +59,157 @@
...
+    
+    2. Anchors are assigned to background when its IoU ratio is lower than
+    negative_overlap (0.4) for all ground-truth boxes.
+    
+    When a anchor is assgned with a ground-truth box which is the i-th category,
+    the i-th entry in its C vector of targets is set to 1 and all other entries
+    are set to 0. When a anchor is assgned with background, all entries are set
+    to 0. Anchors that are not assigned do not contribute to the training
+    objective. The regression targets are the encoded ground-truth boxes
+    associated with the assgined anchors.
```
Comment (jiangjiajun) [inline python/paddle/fluid/layers/detection.py]: 上面这段有三个assigned出错

File: python/paddle/fluid/layers/detection.py [hunk 14]
```diff
@@ -58,6 +59,157 @@
...
+ 
+    Args:
+        bbox_pred(Variable): A 3-D Tensor with shape [N, M, 4] represents the
+            predicted locations of M bounding bboxes. N is the batch size,
+            and each bounding box has four coordinate values and the layout
+            is [xmin, ymin, xmax, ymax].
+        cls_logits(Variable): A 3-D Tensor with shape [N, M, 1] represents the
+            predicted confidence predictions. N is the batch size, 1 is the
+            frontground and background sigmoid, M is number of bounding boxes.
```
Comment (jiangjiajun) [inline python/paddle/fluid/layers/detection.py]: frontground?

File: python/paddle/fluid/layers/detection.py [hunk 15]
```diff
@@ -58,6 +59,157 @@
...
+        anchor_box(Variable): A 2-D Tensor with shape [M, 4] holds M boxes,
+            each box is represented as [xmin, ymin, xmax, ymax],
+            [xmin, ymin] is the left top coordinate of the anchor box,
+            if the input is image feature map, they are close to the origin
+            of the coordinate system. [xmax, ymax] is the right bottom
+            coordinate of the anchor box.
+        anchor_var(Variable): A 2-D Tensor with shape [M,4] holds expanded 
+            variances of anchors.
+        gt_boxes (Variable): The ground-truth boudding boxes (bboxes) are a 2D
```
Comment (jiangjiajun) [inline python/paddle/fluid/layers/detection.py]: boudding?

File: python/paddle/fluid/layers/detection.py [hunk 16]
```diff
@@ -58,6 +59,157 @@
...
+            LoDTensor with shape [Ng, 4], Ng is the total number of ground-truth
+            bboxes of mini-batch input.
+        is_crowd (Variable): A 1-D LoDTensor which indicates groud-truth is crowd.
```
Comment (jiangjiajun) [inline python/paddle/fluid/layers/detection.py]: ground-truth

File: python/paddle/fluid/layers/detection.py [hunk 17]
```diff
@@ -58,6 +59,157 @@
...
+            each box is represented as [xmin, ymin, xmax, ymax],
+            [xmin, ymin] is the left top coordinate of the anchor box,
+            if the input is image feature map, they are close to the origin
+            of the coordinate system. [xmax, ymax] is the right bottom
+            coordinate of the anchor box.
+        anchor_var(Variable): A 2-D Tensor with shape [M,4] holds expanded 
+            variances of anchors.
+        gt_boxes (Variable): The ground-truth boudding boxes (bboxes) are a 2D
+            LoDTensor with shape [Ng, 4], Ng is the total number of ground-truth
+            bboxes of mini-batch input.
```
Comment (jiangjiajun) [inline python/paddle/fluid/layers/detection.py:114]: 每个参数都要添加注释

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 18]
```diff
@@ -573,6 +584,441 @@ negative do not contribute to the training objective.
...
+    AddOutput("BBoxInsideWeight",
+              "(Tensor), The bbox inside weight with shape "
+              "[F, 4], F is the foreground number.");
+    AddOutput("ForegroundNumber",
+              "(Tensor), The foreground number. "
+              "[1, 1].");
+    AddComment(R"DOC(
+    This layer can be, for given the Intersection-over-Union (IoU) overlap
+    between anchors and ground truth boxes, to assign classification and
+    regression targets to each each anchor, these target labels are used for
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: each each anchor?

**Review by qingqing01 — APPROVED** [review]

**Review by xsrobin — APPROVED** [review]
LGTM
